### PR TITLE
69-implement-SURF

### LINKED ISSUE #69: Implement SURF
### The problem
- Get more preference pairs
### The solution / the feature
- Implement SURF 
- https://arxiv.org/pdf/2203.10050
### Which Pros/Cons does this solution have?
- Better performance with less Feedback

### PR DESCRIPTION
**Pull request by Wizzardexe**
**Issue:** Fix #69
- Implement SURF

**What?**
- build option to enable SURF, which enhances feedback efficiency
- cleanup the sampling in main
**How?**
- implemented https://arxiv.org/pdf/2203.10050

**Ready to merge to main?**
- [x] Linter passes
- [ ] PR Approved

### CODE REVIEW

**Comment by Wizzardexe** [comment]
Corrected the labeling issue. The fix should work now, but I haven't been able to fully test it yet.

**Review by jualat — DISMISSED** [review]
This seems to be a appropiate implementation of SURF.

File: framework/reward_net.py [hunk 1]
```diff
@@ -179,11 +179,169 @@ def train_reward(
...
+    unlabeled_batch_ratio=1,
+    tau=0.8,
+    lambda_ssl=0.1,
+    H_max=55,
+    H_min=45,
+):
+    from sampling import sample_pairs
+    from teacher import Preference
+
+    augmentation_factor = (H_max - H_min) / trajectory_length
```
Comment (bkuen) [inline framework/reward_net.py:213]: What's the point of having an augmentation_factor to scale the amount of epochs?

File: framework/reward_net.py [hunk 2]
```diff
@@ -179,11 +179,169 @@ def train_reward(
...
+            sup_loss_accum += ensemble_loss.item()
+
+        unsup_loss_accum = 0.0
+
+        if unlabeled_batch_ratio > 0:
+            unsup_batch_size = unlabeled_batch_ratio * batch_size
+            unlabeled_pairs = sample_pairs(
+                size=unsup_batch_size,
+                rb=rb,
+                sampling_strategy="uniform",
```
Comment (bkuen) [inline framework/reward_net.py:268]: Maybe we should consider different sampling strategies for unlabeled pairs as well?

**Review by bkuen — DISMISSED** [review]
Looks good :) 

Just a few comments for better understanding and maybe further improvement

File: framework/reward_net.py (hunk 2 above)
Comment (Wizzardexe) [reply framework/reward_net.py:268]: Yeah I was going to try this

File: framework/reward_net.py (hunk 1 above)
Comment (Wizzardexe) [reply framework/reward_net.py:213]: I thought it would help to increase the performance of the augmentations, therefore I increase the epochs by the augmentation_factor.

File: framework/sac_rlhf.py [hunk 3]
```diff
@@ -542,39 +544,29 @@ def train(args: Any):
...
+                    trajectory_length=args.trajectory_length,
+                    unlabeled_batch_ratio=args.unlabeled_batch_ratio,
+                    tau=args.surf_tau,
+                    lambda_ssl=args.lambda_ssl,
+                    H_max=args.surf_H_max,
+                    H_min=args.surf_H_min,
+                )
                 rb.relabel_rewards(reward_net)
+                train_pref_buffer.reset()
+                val_pref_buffer.reset()
```
Comment (Wizzardexe) [inline framework/sac_rlhf.py:569]: After the training we now reset the buffer.

File: framework/reward_net.py [hunk 4]
```diff
@@ -236,13 +257,73 @@ def train_or_val_pref_batch(
             torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
             optimizer.step()
 
-        total_loss += ensemble_loss.item()
+        sup_loss_accum += ensemble_loss.item()
+    sup_avg_loss = sup_loss_accum / len(prefs)
+
+    unsup_loss_accum = 0.0
+    unsup_avg_loss = 0.0
```
Comment (Wizzardexe) [inline framework/reward_net.py:267]: @bkuen do you think we should include this loss in the validation aswell

File: framework/reward_net.py [hunk 5]
```diff
@@ -192,35 +200,48 @@ def train_or_val_pref_batch(
...
+    sup_loss_accum = 0.0
     for pref_pair in prefs:
         t1_start_idx, t1_end_idx, t2_start_idx, t2_end_idx, pref = pref_pair
         pref = torch.tensor(pref, dtype=torch.float32).to(device)
 
         t1 = rb.get_trajectory(int(t1_start_idx), int(t1_end_idx), env=env)
         t2 = rb.get_trajectory(int(t2_start_idx), int(t2_end_idx), env=env)
 
+        t1_aug = rb.temporal_data_augmentation(t1, H_max=H_max, H_min=H_min, env=env)
+        t2_aug = rb.temporal_data_augmentation(t2, H_max=H_max, H_min=H_min, env=env)
```
Comment (Wizzardexe) [inline framework/reward_net.py:232]: Rightnow we always use augmentation, I could also change this to only use augmentation when using surf

File: framework/sac_rlhf.py [hunk 6]
```diff
@@ -89,7 +89,7 @@ class Args:
     ## Evaluation
     evaluation_frequency: int = 10000
     """the frequency of evaluation"""
-    evaluation_episodes: int = 30
+    evaluation_episodes: int = 10
```
Comment (Wizzardexe) [inline framework/sac_rlhf.py:92]: 10 episodes should be enough for evaluation while training, decreasing the amount should speed up the training.

File: framework/reward_net.py (hunk 5 above)
Comment (bkuen) [reply framework/reward_net.py:232]: I think we should only use augmentation when using surf because the whole purpose of surf is augmentation. Always using temporal augmentation makes the surf flag kinda obsolete, I like it when all the additional customizations to the Pebble algorithm can be activated by the flag.

File: framework/reward_net.py (hunk 4 above)
Comment (bkuen) [reply framework/reward_net.py:267]: I think, it's totally fine to include the unsupervised loss in the validation. Thus, I would suggest that we return a dictionary with `total avg loss`, `sup avg loss`and `unsup avg loss`, and log all values.

**Review by bkuen — CHANGES_REQUESTED** [review]

**Review by bkuen — APPROVED** [review]
